include/boost/url/detail/format_args.hpp

```diff
@@ -0,0 +1,1055 @@
+    {
+        // get n digits
+        T v0 = v;
+        T p = 1;
+        std::size_t n = 0;
+        if (is_negative(v))
+        {
+            invert(v);
+            ++n;
+        }
+        else if (sign != '-')
+        {
+            ++n;
+        }
+        while (v > 0)
+        {
+            if (v >= 10)
+                p *= 10;
+            v /= 10;
+            ++n;
+        }
+        static constexpr auto m =
+            std::numeric_limits<T>::digits10;
+        BOOST_ASSERT(n <= m + 1);
+        ignore_unused(m);
+
+        // get pad
+        std::size_t w = width;
+        if (width_idx != std::size_t(-1) ||
+            !width_name.empty())
+        {
+            get_width_from_args(
+                width_idx, width_name, ctx.args(), w);
+        }
+        std::size_t lpad = 0;
+        std::size_t rpad = 0;
+        if (w > n)
+        {
+            std::size_t pad = w - n;
+            if (zeros)
+            {
+                lpad = pad;
+            }
+            else
+            {
+                switch (align)
+                {
+                case '<':
+                    rpad = pad;
+                    break;
+                case '>':
+                    lpad = pad;
+                    break;
+                case '^':
+                    lpad = pad / 2;
+                    rpad = pad - lpad;
+                    break;
+                }
+            }
+        }
+
+        // write
+        v = v0;
+        char* out = ctx.out();
+        if (!zeros)
+        {
+            for (std::size_t i = 0; i < lpad; ++i)
+                encode_one(out, fill, cs);
+        }
+        if (is_negative(v))
+        {
+            encode_one(out, '-', cs);
+            invert(v);
+            --n;
+        }
+        else if (sign != '-')
+        {
+            encode_one(out, sign, cs);
+            --n;
+        }
+        if (zeros)
+        {
+            for (std::size_t i = 0; i < lpad; ++i)
+                encode_one(out, '0', cs);
+        }
+        while (n)
+        {
+            unsigned long long int d = v / p;
+            encode_one(out, '0' + static_cast<char>(d), cs);
+            --n;
+            v %= p;
+            p /= 10;
+        }
+        if (!zeros)
+        {
+            for (std::size_t i = 0; i < rpad; ++i)
+                encode_one(out, fill, cs);
+        }
+        return out;
+    }
+
+    static
+    bool
+    is_negative(T v)
+    {
+        return is_negative(v, std::is_signed<T>{});
+    }
+
+    static
+    bool
+    is_negative(T v, std::true_type)
+    {
+        return v < 0;
+    }
+
+    static
+    bool
+    is_negative(T, std::false_type)
+    {
+        return false;
+    }
+
+    static
+    void
+    invert(T& v)
+    {
+        invert(v, std::is_signed<T>{});
+    }
+
+    static
+    void
+    invert(T& v, std::true_type)
+    {
+        v *= -1;
+    }
+
+    static
+    void
+    invert(T&, std::false_type)
+    {}
+};
```

Comment:
1000+ line header file, is this visible to users? Or is it only included from a .ipp file?

Reply:
Most content has been moved to ipp files now.